OgreMain/src/OgreParticleSystem.cpp

```diff
@@ -549,62 +500,37 @@ namespace Ogre {
     void ParticleSystem::_executeTriggerEmitters(ParticleEmitter* emitter, unsigned requested, Real timeElapsed)
     {
         Real timePoint = 0.0f;
-
-
-        // avoid any divide by zero conditions
-        if(!requested) 
+        if(!requested)
             return;
-
         Real timeInc = timeElapsed / requested;
-
-        for (unsigned int j = 0; j < requested; ++j)
+        unsigned start = 0;
+        unsigned end = 0;
+        mParticles.create (requested, start, end);
+        emitter->_initParticles (mParticles, start, end);
+        if (!mLocalSpace)
         {
-            // Create a new particle & init using emitter
-            // The particle is a visual particle if the emit_emitter property of the emitter isn't set 
-            Particle* p = 0;
-            String  emitterName = emitter->getEmittedEmitter();
-            if (emitterName.empty())
-                p = createParticle();
-            else
-                p = createEmitterParticle(emitterName);
```

Comment:
this prevents you from creating "Fireworks"-style systems (as in the ParticleFX sample).

Reply:
i am planning to add the emitter emitters back in -- this was just far enough to test out the standard perf.

Reply:
a better design for emitting emitters is to create a new emitter (say EmitEmitter) in fx rather than include this code in OgreParticleSystem in OgreMain.  if you pollute the OgreParticleSystem class with "emitting emitters" you add unnecessary complexity and decrease the performance of all particle types even when you're not using emitting emitters.  you should only take the hit for emitting emitters if you're using them.

Reply:
ogre needs a "technique" concept like particle universe (which is feature rich but too complicated to refactor).  internally i'm calling mine an "effect"... which allows a script to include and combine multiple ParticleSystems into a single visual effect.  

this allows you to mix techniques and not require the core particle design to be overly complicated.

for example, rather than allowing individual particles to be of any variety of sizes (which bloats memory, adds more code paths per frame, etc), you just create an "effect" which is comprised of multiple ParticleSystems that each use their own size... so you're only paying for what you use and not punishing the particle system even when not using this feature.

Reply:
also note the +30% perf improvement is on a single particle system with a single emitter and no affectors.  here's torchlight 2's blood splatter particle effect:  (blood shoots everywhere, chunks bounce off the ground, etc...)

![image](https://user-images.githubusercontent.com/976243/95705818-02233480-0c0a-11eb-8da0-f103dc7513a0.png)

each line in this "particle" is a single particle system.  

here is the same effect expanded but not showing everything:

![image](https://user-images.githubusercontent.com/976243/95705874-2da61f00-0c0a-11eb-916c-8fb1a406af49.png)

imagine the savings for each as all these affectors again and again and again iterate through the linked lists of large old school Ogre::Particle structs which are huge.  the cache is trashed over and over and over.

Reply:
> if you pollute the OgreParticleSystem class with "emitting emitters" you add unnecessary complexity and decrease the performance of all particle types even when you're not using emitting emitters.

yes, the current implementation is bad. However, I think if you let observe emitters a particle for positioning instead of making them one, you can make "emitted emitters" free if you do not use them, while still having them in the API.

> for example, rather than allowing individual particles to be of any variety of sizes 

individual particle size just takes 50% of individual particle colour or +16% of (position + velocity + colour + ttl) and it allows non-uniform scale affectors. I would not say it is that bad.

> imagine the savings for each as all these affectors again and again and again

I get what you are aiming at, but it is still 30% :wink:

Reply:
got the perf up to a 50%+ improvement.  i had to copy/paste Ogre::BillboardSet and manually "unroll" the injectBillboard() function because it's being called now in a tight loop and it frequently accesses member variables.  the members do not appear to vary across the tight loop but i imagine the compiler can't make that assumption.  i wound up pulling up all the "options" and query them up front, then pass down pointers to member variables based on the decision.

i'll stop here... i'm happy with a 2x particle budget increase.  

thanks for all the help!

Reply:
for example, i first check to see how the vertexes should be created:
        void (Ogre::BillboardSetFx::*genVerticesFn)(const Ogre::Vector3* const offsets, const Ogre::Billboard & pBillboard) = nullptr;
        auto vertexPolicy = mBillboardSet->getVertexPolicy ();
        if (vertexPolicy == Ogre::BillboardSetFx::rotation)
            genVerticesFn = &Ogre::BillboardSetFx::genVerticesRotation;
        else if (vertexPolicy == Ogre::BillboardSetFx::vertex)
            genVerticesFn = &Ogre::BillboardSetFx::genVerticesVertex;
        else if (vertexPolicy == Ogre::BillboardSetFx::point)
            genVerticesFn = &Ogre::BillboardSetFx::genVerticesPoint;
and then pass this function down into injectBillboard().

Reply:
it's ugly but sometimes i'm getting close to 60fps again with 100k particles @4k resolution.

Reply:
lastly, in my runtime i'll allocate the particle systems in the same block of memory as the affectors and emitters.